// @azure-tools/typespec-azure-core/dist/src/rules/prevent-format.js
import { createRule } from "@typespec/compiler";
export const preventFormatRule = createRule({
    name: "no-format",
    description: "Azure services should not use the `@format` decorator.",
    severity: "warning",
    url: "https://azure.github.io/typespec-azure/docs/libraries/azure-core/rules/prevent-format",
    messages: {
        default: "Azure services should not use the `@format` decorator.",
    },
    create(context) {
        return {
            scalar: (scalar) => {
                for (const dec of scalar.decorators) {
                    if (dec.decorator.name === "$format") {
                        context.reportDiagnostic({
                            target: dec.node ? dec.node : scalar,
                        });
                    }
                }
            },
            modelProperty: (model) => {
                for (const dec of model.decorators) {
                    if (dec.decorator.name === "$format") {
                        context.reportDiagnostic({
                            target: dec.node ? dec.node : model,
                        });
                    }
                }
            },
        };
    },
});
