// @typespec/compiler/dist/src/core/helpers/projected-names-utils.js
import { projectProgram } from "../program.js";
/**
 * Create an helper to manager project names.
 * @param program Program
 * @param target Name of the projected name target(e.g. json, csharp, etc.)
 * @returns ProjectedNameView
 */
export function createProjectedNameProgram(program, target) {
    const projectedProgram = projectProgram(program, [
        {
            projectionName: "target",
            arguments: [target],
        },
    ]);
    return {
        program: projectedProgram,
        getProjectedName,
    };
    function getProjectedName(type) {
        const baseType = findTypeInProjector(projectedProgram.projector, type);
        const projectedType = projectedProgram.projector.projectedTypes.get(baseType);
        if (projectedType === undefined ||
            !("name" in projectedType) ||
            projectedType.name === baseType.name) {
            return type.name;
        }
        return projectedType.name;
    }
}
// TODO will REALM help here?
function findTypeInProjector(projector, type) {
    if (type.projectionSource === undefined) {
        return type;
    }
    else if (type.projector === projector) {
        return type;
    }
    else {
        return findTypeInProjector(projector, type.projectionSource);
    }
}
