// @azure-tools/typespec-azure-resource-manager/dist/src/rules/arm-resource-invalid-action-verb.js
import { createRule } from "@typespec/compiler";
import { getOperationVerb } from "@typespec/http";
import { getActionDetails } from "@typespec/rest";
import { isInternalTypeSpec } from "./utils.js";
export const invalidActionVerbRule = createRule({
    name: "arm-resource-invalid-action-verb",
    severity: "warning",
    description: "Actions must be HTTP Post operations.",
    messages: {
        default: "Actions must be HTTP Post operations.",
    },
    create(context) {
        return {
            operation: (operation) => {
                if (!isInternalTypeSpec(context.program, operation)) {
                    const actionType = getActionDetails(context.program, operation);
                    const verb = getOperationVerb(context.program, operation);
                    if (actionType !== undefined && verb !== "post") {
                        context.reportDiagnostic({
                            target: operation,
                        });
                    }
                }
            },
        };
    },
});
