// @azure-tools/typespec-java/dist/src/common/schemas/time.js
/* eslint-disable @typescript-eslint/no-unsafe-declaration-merging */
import { PrimitiveSchema, SchemaType } from "@autorest/codemodel";
export class DurationSchema extends PrimitiveSchema {
    constructor(name, description, objectInitializer) {
        super(name, description, SchemaType.Duration);
        this.format = "duration-rfc3339";
        this.apply(objectInitializer);
    }
}
export class DateTimeSchema extends PrimitiveSchema {
    constructor(name, description, objectInitializer) {
        super(name, description, SchemaType.DateTime);
        this.apply(objectInitializer);
    }
}
export class DateSchema extends PrimitiveSchema {
    constructor(name, description, objectInitializer) {
        super(name, description, SchemaType.Date);
        this.apply(objectInitializer);
    }
}
export class TimeSchema extends PrimitiveSchema {
    constructor(name, description, objectInitializer) {
        super(name, description, SchemaType.Time);
        this.apply(objectInitializer);
    }
}
export class UnixTimeSchema extends PrimitiveSchema {
    constructor(name, description, objectInitializer) {
        super(name, description, SchemaType.UnixTime);
        this.apply(objectInitializer);
    }
}
