// @azure-tools/typespec-java/dist/src/common/schemas/usage.js
export var SchemaContext;
(function (SchemaContext) {
    /** Schema is used as an input to an operation. */
    SchemaContext["Input"] = "input";
    /** Schema is used as an output from an operation. */
    SchemaContext["Output"] = "output";
    /** Schema is used as an exception from an operation. */
    SchemaContext["Exception"] = "exception";
    /** Schema is used from the operation for generating convenience API. And it is public */
    SchemaContext["Public"] = "public";
    /** Schema is used for internal API, or dev purpose. Not exposed to user. */
    SchemaContext["Internal"] = "internal";
    /** Schema is used from the pageable operation. This usage does not propagate. */
    SchemaContext["Paged"] = "paged";
    /** Schema as anonymous model. This usage does not propagate. */
    SchemaContext["Anonymous"] = "anonymous";
    /** Schema is used in json-merge-patch operation */
    SchemaContext["JsonMergePatch"] = "json-merge-patch";
})(SchemaContext || (SchemaContext = {}));
