From ee9efb32a58711f5bcb4667e4ae9bea70be2e138 Mon Sep 17 00:00:00 2001
From: Weidong Xu <weidxu@microsoft.com>
Date: Thu, 7 Aug 2025 13:47:27 +0800
Subject: [PATCH] revert impl in easm

---
 .../easm/implementation/EasmClientImpl.java   | 30 +++++++++++--------
 1 file changed, 18 insertions(+), 12 deletions(-)

diff --git a/sdk/easm/azure-analytics-defender-easm/src/main/java/com/azure/analytics/defender/easm/implementation/EasmClientImpl.java b/sdk/easm/azure-analytics-defender-easm/src/main/java/com/azure/analytics/defender/easm/implementation/EasmClientImpl.java
index 74965f145fd..cd574da24ef 100644
--- a/sdk/easm/azure-analytics-defender-easm/src/main/java/com/azure/analytics/defender/easm/implementation/EasmClientImpl.java
+++ b/sdk/easm/azure-analytics-defender-easm/src/main/java/com/azure/analytics/defender/easm/implementation/EasmClientImpl.java
@@ -315,7 +315,7 @@ public final class EasmClientImpl {
         @UnexpectedResponseExceptionType(HttpResponseException.class)
         Mono<Response<Void>> deleteDataConnection(@HostParam("endpoint") String endpoint,
             @QueryParam("api-version") String apiVersion, @PathParam("dataConnectionName") String dataConnectionName,
-            RequestOptions requestOptions, Context context);
+            @HeaderParam("Accept") String accept, RequestOptions requestOptions, Context context);
 
         @Delete("/dataConnections/{dataConnectionName}")
         @ExpectedResponses({ 204 })
@@ -325,7 +325,7 @@ public final class EasmClientImpl {
         @UnexpectedResponseExceptionType(HttpResponseException.class)
         Response<Void> deleteDataConnectionSync(@HostParam("endpoint") String endpoint,
             @QueryParam("api-version") String apiVersion, @PathParam("dataConnectionName") String dataConnectionName,
-            RequestOptions requestOptions, Context context);
+            @HeaderParam("Accept") String accept, RequestOptions requestOptions, Context context);
 
         @Get("/discoGroups")
         @ExpectedResponses({ 200 })
@@ -419,7 +419,7 @@ public final class EasmClientImpl {
         @UnexpectedResponseExceptionType(HttpResponseException.class)
         Mono<Response<Void>> runDiscoGroup(@HostParam("endpoint") String endpoint,
             @QueryParam("api-version") String apiVersion, @PathParam("groupName") String groupName,
-            RequestOptions requestOptions, Context context);
+            @HeaderParam("Accept") String accept, RequestOptions requestOptions, Context context);
 
         @Post("/discoGroups/{groupName}:run")
         @ExpectedResponses({ 204 })
@@ -429,7 +429,7 @@ public final class EasmClientImpl {
         @UnexpectedResponseExceptionType(HttpResponseException.class)
         Response<Void> runDiscoGroupSync(@HostParam("endpoint") String endpoint,
             @QueryParam("api-version") String apiVersion, @PathParam("groupName") String groupName,
-            RequestOptions requestOptions, Context context);
+            @HeaderParam("Accept") String accept, RequestOptions requestOptions, Context context);
 
         @Get("/discoGroups/{groupName}/runs")
         @ExpectedResponses({ 200 })
@@ -625,7 +625,7 @@ public final class EasmClientImpl {
         @UnexpectedResponseExceptionType(HttpResponseException.class)
         Mono<Response<Void>> deleteSavedFilter(@HostParam("endpoint") String endpoint,
             @QueryParam("api-version") String apiVersion, @PathParam("filterName") String filterName,
-            RequestOptions requestOptions, Context context);
+            @HeaderParam("Accept") String accept, RequestOptions requestOptions, Context context);
 
         @Delete("/savedFilters/{filterName}")
         @ExpectedResponses({ 204 })
@@ -635,7 +635,7 @@ public final class EasmClientImpl {
         @UnexpectedResponseExceptionType(HttpResponseException.class)
         Response<Void> deleteSavedFilterSync(@HostParam("endpoint") String endpoint,
             @QueryParam("api-version") String apiVersion, @PathParam("filterName") String filterName,
-            RequestOptions requestOptions, Context context);
+            @HeaderParam("Accept") String accept, RequestOptions requestOptions, Context context);
 
         @Get("/tasks")
         @ExpectedResponses({ 200 })
@@ -1886,8 +1886,9 @@ public final class EasmClientImpl {
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Response<Void>> deleteDataConnectionWithResponseAsync(String dataConnectionName,
         RequestOptions requestOptions) {
+        final String accept = "application/json";
         return FluxUtil.withContext(context -> service.deleteDataConnection(this.getEndpoint(),
-            this.getServiceVersion().getVersion(), dataConnectionName, requestOptions, context));
+            this.getServiceVersion().getVersion(), dataConnectionName, accept, requestOptions, context));
     }
 
     /**
@@ -1903,8 +1904,9 @@ public final class EasmClientImpl {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Response<Void> deleteDataConnectionWithResponse(String dataConnectionName, RequestOptions requestOptions) {
+        final String accept = "application/json";
         return service.deleteDataConnectionSync(this.getEndpoint(), this.getServiceVersion().getVersion(),
-            dataConnectionName, requestOptions, Context.NONE);
+            dataConnectionName, accept, requestOptions, Context.NONE);
     }
 
     /**
@@ -2696,8 +2698,9 @@ public final class EasmClientImpl {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Response<Void>> runDiscoGroupWithResponseAsync(String groupName, RequestOptions requestOptions) {
+        final String accept = "application/json";
         return FluxUtil.withContext(context -> service.runDiscoGroup(this.getEndpoint(),
-            this.getServiceVersion().getVersion(), groupName, requestOptions, context));
+            this.getServiceVersion().getVersion(), groupName, accept, requestOptions, context));
     }
 
     /**
@@ -2713,7 +2716,8 @@ public final class EasmClientImpl {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Response<Void> runDiscoGroupWithResponse(String groupName, RequestOptions requestOptions) {
-        return service.runDiscoGroupSync(this.getEndpoint(), this.getServiceVersion().getVersion(), groupName,
+        final String accept = "application/json";
+        return service.runDiscoGroupSync(this.getEndpoint(), this.getServiceVersion().getVersion(), groupName, accept,
             requestOptions, Context.NONE);
     }
 
@@ -4053,8 +4057,9 @@ public final class EasmClientImpl {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Response<Void>> deleteSavedFilterWithResponseAsync(String filterName, RequestOptions requestOptions) {
+        final String accept = "application/json";
         return FluxUtil.withContext(context -> service.deleteSavedFilter(this.getEndpoint(),
-            this.getServiceVersion().getVersion(), filterName, requestOptions, context));
+            this.getServiceVersion().getVersion(), filterName, accept, requestOptions, context));
     }
 
     /**
@@ -4070,8 +4075,9 @@ public final class EasmClientImpl {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Response<Void> deleteSavedFilterWithResponse(String filterName, RequestOptions requestOptions) {
+        final String accept = "application/json";
         return service.deleteSavedFilterSync(this.getEndpoint(), this.getServiceVersion().getVersion(), filterName,
-            requestOptions, Context.NONE);
+            accept, requestOptions, Context.NONE);
     }
 
     /**
-- 
2.50.1.windows.1

