From 4a93f9b65224b9c8dafc856f6de43d876e5c40a3 Mon Sep 17 00:00:00 2001
From: Weidong Xu <weidxu@microsoft.com>
Date: Mon, 22 Sep 2025 15:32:34 +0800
Subject: [PATCH] revert Impl for onlineexperimentation

---
 .../OnlineExperimentationClientImpl.java               | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/sdk/onlineexperimentation/azure-analytics-onlineexperimentation/src/main/java/com/azure/analytics/onlineexperimentation/implementation/OnlineExperimentationClientImpl.java b/sdk/onlineexperimentation/azure-analytics-onlineexperimentation/src/main/java/com/azure/analytics/onlineexperimentation/implementation/OnlineExperimentationClientImpl.java
index 32e6622cbc6..ed26312e71d 100644
--- a/sdk/onlineexperimentation/azure-analytics-onlineexperimentation/src/main/java/com/azure/analytics/onlineexperimentation/implementation/OnlineExperimentationClientImpl.java
+++ b/sdk/onlineexperimentation/azure-analytics-onlineexperimentation/src/main/java/com/azure/analytics/onlineexperimentation/implementation/OnlineExperimentationClientImpl.java
@@ -233,7 +233,7 @@ public final class OnlineExperimentationClientImpl {
         @UnexpectedResponseExceptionType(HttpResponseException.class)
         Mono<Response<Void>> deleteMetric(@HostParam("endpoint") String endpoint,
             @QueryParam("api-version") String apiVersion, @PathParam("experimentMetricId") String experimentMetricId,
-            RequestOptions requestOptions, Context context);
+            @HeaderParam("Accept") String accept, RequestOptions requestOptions, Context context);
 
         @Delete("/experiment-metrics/{experimentMetricId}")
         @ExpectedResponses({ 204 })
@@ -243,7 +243,7 @@ public final class OnlineExperimentationClientImpl {
         @UnexpectedResponseExceptionType(HttpResponseException.class)
         Response<Void> deleteMetricSync(@HostParam("endpoint") String endpoint,
             @QueryParam("api-version") String apiVersion, @PathParam("experimentMetricId") String experimentMetricId,
-            RequestOptions requestOptions, Context context);
+            @HeaderParam("Accept") String accept, RequestOptions requestOptions, Context context);
 
         @Get("/experiment-metrics")
         @ExpectedResponses({ 200 })
@@ -698,8 +698,9 @@ public final class OnlineExperimentationClientImpl {
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Response<Void>> deleteMetricWithResponseAsync(String experimentMetricId,
         RequestOptions requestOptions) {
+        final String accept = "application/json";
         return FluxUtil.withContext(context -> service.deleteMetric(this.getEndpoint(),
-            this.getServiceVersion().getVersion(), experimentMetricId, requestOptions, context));
+            this.getServiceVersion().getVersion(), experimentMetricId, accept, requestOptions, context));
     }
 
     /**
@@ -730,8 +731,9 @@ public final class OnlineExperimentationClientImpl {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Response<Void> deleteMetricWithResponse(String experimentMetricId, RequestOptions requestOptions) {
+        final String accept = "application/json";
         return service.deleteMetricSync(this.getEndpoint(), this.getServiceVersion().getVersion(), experimentMetricId,
-            requestOptions, Context.NONE);
+            accept, requestOptions, Context.NONE);
     }
 
     /**
-- 
2.51.0.windows.1

