// @typespec/compiler/dist/src/init/file-templating.js
import { camelCase, kebabCase, pascalCase } from "change-case";
import Mustache from "mustache";
export function createFileTemplatingContext(config) {
    return {
        ...config.template,
        ...config,
        normalizeVersion,
        toLowerCase,
        normalizePackageName,
        casing,
    };
}
export function render(content, context) {
    return Mustache.render(content, context);
}
const casing = {
    camelCase: createNamingUtils(camelCase),
    kebabCase: createNamingUtils(kebabCase),
    pascalCase: createNamingUtils(pascalCase),
};
function createNamingUtils(fn) {
    return () => (text, render) => fn(render(text));
}
const normalizeVersion = function () {
    return function (text, render) {
        return render(text).replaceAll("-", "_");
    };
};
const toLowerCase = function () {
    return function (text, render) {
        return render(text).toLowerCase();
    };
};
const normalizePackageName = function () {
    return function (text, render) {
        return render(text).replaceAll(".", "-").toLowerCase();
    };
};
