// @typespec/compiler/dist/src/init/init-template.js
import { TypeSpecConfigJsonSchema } from "../config/config-schema.js";
export const InitTemplateLibrarySpecSchema = {
    type: "object",
    additionalProperties: false,
    properties: {
        name: { type: "string" },
        version: { type: "string", nullable: true },
    },
    required: ["name"],
};
export const InitTemplateSchema = {
    type: "object",
    additionalProperties: false,
    properties: {
        title: { type: "string" },
        description: { type: "string" },
        compilerVersion: { type: "string", nullable: true },
        libraries: {
            type: "array",
            items: {
                oneOf: [{ type: "string" }, InitTemplateLibrarySpecSchema],
            },
            nullable: true,
        },
        skipCompilerPackage: { type: "boolean", nullable: true },
        config: { nullable: true, ...TypeSpecConfigJsonSchema },
        inputs: {
            type: "object",
            nullable: true,
            additionalProperties: {
                type: "object",
                properties: {
                    description: { type: "string" },
                    type: { type: "string", enum: ["text"] },
                    initialValue: {},
                },
                required: ["description", "type"],
            },
            required: [],
        },
        files: {
            type: "array",
            nullable: true,
            items: {
                type: "object",
                properties: {
                    path: { type: "string" },
                    destination: { type: "string" },
                    skipGeneration: { type: "boolean", nullable: true },
                },
                required: ["path", "destination"],
            },
        },
    },
    required: ["title", "description"],
};
